### PR TITLE
Add user-callable functions to plugins

### PR DESCRIPTION
**Pull request by Bensk1**
As highlighted by @Bouncner in #2437 there are many scenarios where the user (_Anwender_) wants to trigger a plugin function, e.g., to manually start reoptimization. 

There are two challenges to solve for this to work.

1. How should the user interaction look like? 
1. How can the PluginManager call plugin functions that are not known in advance without modifying the PluginManager. 

A possible solution for challenge 1 could be to use a new meta table. Inserting a row containing the name of the plugin and action to execute into this table could trigger the action.

This is the first proposal for a solution for challenge 2. The plugin's developer can just implement user-callable functions as a _normal_ plugin function. Afterward, the developer only has to use the macro `EXPORT_USER_CALLABLE_FUNCTION` that the AbstractPlugin provides. 

Alternatively, @Bouncner also had the idea that plugins can subscribe keywords and function calls to the PluginManager upon the instantiation of the plugins. The PluginManager maintains a list of such keywords. If a keyword is sent, the corresponding function is called. This might be a cleaner solution but also require more changes and I don't find the above's solution particularly unclean.
~~_Edit: I don't know whether this solution would work at all. The PluginManager would need to keep a map of keywords and function pointers. However, function pointers to member functions also contain the class type. For different plugins, the class type varies so we could not create such a map at compile time. Using AbstractPlugin as type would also not work because then we would need to define all user-callable functions for AbstractPlugin, too. Am I missing something or would this really not work?_~~ << It [does work,](https://gist.github.com/Bouncner/6d31c7ebe55aeb9d9fb8d0a5efb654fb) as @Bouncner pointed out.

The interface via MetaTables would be mostly identical for both solutions. Also, I believe both solutions can only offer non-parameterized and void-returning functions as the types must be known during the PluginManager's compile time.

### CODE REVIEW

File: src/lib/utils/abstract_plugin.hpp [hunk 1]
```diff
@@ -12,6 +12,21 @@ namespace opossum {
...
+// This macro allows the function `FunctionName` of the plugin called `PluginName` to be
+// externally callable. This is necessary to enable user-callable functions. As above,
+// it leads to plain C linkage. Use EXPORT in the implementation file of your plugin.
+#define EXPORT_USER_CALLABLE_FUNCTION(PluginName, FunctionName) \
+  /* Careful! We are getting a raw pointer from a smart (unique) pointer here. */ \
+  /* This could conflict with the intention of unique_ptr. It should not be    */ \
+  /* a problem, as long as we handle it responsible and let it go out of scope */ \
+  /* at the end of this function.                                              */ \
+  extern "C" void FunctionName ## _extern(const std::unique_ptr<AbstractPlugin>& plugin) { \
+    const auto plugin_raw = plugin.get(); \
```
Comment (Bensk1) [inline src/lib/utils/abstract_plugin.hpp:24]: Unfortunately, I think there is no way around that.

**Comment by Bouncner** [comment]
> The interface via MetaTables would be mostly identical for both solutions. Also, I believe both solutions can only offer non-parameterized and void-returning functions as the types must be known during the PluginManager's compile time.

Yes, I don't see a nice way around non-parameterized void functions. But that shouldn't be a big issue. Return data can be written to the meta log and parameters can be obtained via the settings.

**Comment by Bensk1** [comment]
> Yes, I don't see a nice way around non-parameterized void functions. But that shouldn't be a big issue. Return data can be written to the meta log and parameters can be obtained via the settings.

Exactly.

**Comment by Bensk1** [comment]
[This is](https://github.com/hyrise/hyrise/commit/430679a2021d2b8455b90158bf0d57dcc04b6427) how the alternative implementation would look like.

File: src/lib/utils/abstract_plugin.hpp (hunk 1 above)
Comment (mweisgut) [reply src/lib/utils/abstract_plugin.hpp:25]: Do we have to use an `std::unique_ptr` here?

**Comment by mweisgut** [comment]
From a user experience perspective, I personally would prefer executing the SQL command `PRAGMA test_user_callable_function` instead of `INSERT INTO meta_exec (plugin_name, function_name) VALUES ('hyriseTestPlugin', 'test_user_callable_function_extern')`.

For this, a pragma operator could be added, which calls the plugins' function calls and could also allow enabling/loading and disabling/unloading plugins.
Furthermore, in [this approach](https://github.com/hyrise/hyrise/commit/430679a2021d2b8455b90158bf0d57dcc04b6427), another way is to map a keyword instead of a {plugin name, keyword} pair to an std::function. This, however, requires that the keywords are unique across all plugins, which can be ensured by naming guidelines and integration tests.

Nevertheless, the `IMPORT` approach probably requires less effort since it does not require an additional operator.

>Yes, I don't see a nice way around non-parameterized void functions.

I think it would not be assessed as `nice way around`, but string-only functions might be a potential way to provide parameterized functions. In the function body, the parameters can then be converted to the required data type.

**Comment by Bensk1** [comment]
> From a user experience perspective, I personally would prefer executing the SQL command `PRAGMA test_user_callable_function` instead of `INSERT INTO meta_exec (plugin_name, function_name) VALUES ('hyriseTestPlugin', 'test_user_callable_function_extern')`.

I agree, it might be a bit cleaner from a user perspective. On the other hand, `PRAGMA` is not part of the SQL standard, is it? Also we would need a new operator and add non-standard keywords to the parser just for a feature that is rarely used. Lastly, in industry it is apparently quite common to use special tables (similar to what we do with meta tables) for such tasks. I am not against the PRAGMA but I believe the other approach is more PRAGMAtic (haha).

> For this, a pragma operator could be added, which calls the plugins' function calls and could also allow enabling/loading and disabling/unloading plugins. Furthermore, in [this approach](https://github.com/hyrise/hyrise/commit/430679a2021d2b8455b90158bf0d57dcc04b6427), another way is to map a keyword instead of a {plugin name, keyword} pair to an std::function. This, however, requires that the keywords are unique across all plugins, which can be ensured by naming guidelines and integration tests.

So what's the advantage of just using a keyword instead of the pair? I just see additional limitations: only unique keywords. Also, since the development plugins is on purpose decoupled from the core, I don't know if we could easily build such integration tests. 

> Nevertheless, the `IMPORT` approach probably requires less effort since it does not require an additional operator.

Just to be sure, you meant `INSERT` here, right?

> I think it would not be assessed as `nice way around`, but string-only functions might be a potential way to provide parameterized functions. In a corresponding function body, the parameters can then be converted to the required data types.

Good point, that could be an option. For now, I find the current implementation sufficient as we can use the settings table for parameters as @Bouncner suggested.

File: src/lib/utils/abstract_plugin.hpp (hunk 1 above)
Comment (Bensk1) [reply src/lib/utils/abstract_plugin.hpp:25]: The PluginManager stores plugins as unique_ptrs. I think in the beginning we thought that the PluginManager should have exclusive ownership and that there is no reason for shared_ptrs. We could change that but might go with the alternative implementation in the end anyway.

**Comment by mweisgut** [comment]
> So what's the advantage of just using a keyword instead of the pair?

I think convenience on the user side by simplified/shorter SQL commands is the only benefit.

> Just to be sure, you meant INSERT here, right?

Right.

File: src/lib/utils/abstract_plugin.hpp (hunk 1 above)
Comment (mweisgut) [reply src/lib/utils/abstract_plugin.hpp:25]: With shared pointers, we can avoid a raw pointer object in this code location, a raw pointer cast, and the comments explaining why we use a raw pointer here.
The minimal overhead for reference counting with shared pointers should also be negligible. Am I missing some drawbacks of changing to shared pointers here?

**Comment by Bouncner** [comment]
Hmm, seeing this unique_ptr “issue”, I’d vote for the functor.
Especially with the `register_keyword()` function, the interface is much better, no extern usage, and same functionality.

**Comment by mweisgut** [comment]
> Hmm, seeing this unique_ptr “issue”, I’d vote for the functor map solution. Especially with the `register_keyword()` function, the interface is much better, no extern usage, and same functionality.

+1

File: src/lib/utils/abstract_plugin.hpp (hunk 1 above)
Comment (Bensk1) [reply src/lib/utils/abstract_plugin.hpp:25]: Technically, we could change to a `shared_ptr` but we chose the `unique_ptr` on purpose to declare exclusive ownership. The idea was that the PluginManager is not sharing plugin ownership because otherwise things could get nasty when unloading the shared library.

**Comment by Bensk1** [comment]
> Hmm, seeing this unique_ptr “issue”, I’d vote for the functor map solution.
> Especially with the `register_keyword()` function, the interface is much better, no extern usage, and same functionality.

Just for clarification, the `extern` only enables C linkage to avoid name mangling. We use to be able to call `test_user_callable_function` instead of `_ZN9Opossum7TestPlugin6test_user_callable_functionEv`.

What I find the biggest argument for the alternative map solution: it enables us to allow providing a list of all available execution functions such that the plugin user does not have to consult the plugin's documentation for that.

**Comment by Bouncner** [comment]
I read a little bit about extern and yes, it's apparently no magic. But it's still kind of a work around from a pure C++ perspective and I'd say we need a very good reason to use workarounds.

So the advantages I see are (i) pure C++ with easy debugging (e.g., looping prints over the map) and (ii) the explicit `register_keyword()` API. But both are not strong points. Unfortunately, I am not sufficiently familiar with this whole topic to estimate potential problems that might occur later on (anything to be aware of when dynamic loading magic is used? I don't know!)

**Comment by dey4ss** [comment]
I like the idea of calling plugin code from outside. For instance, it would be a nice functionality for my master's thesis plugin. Regarding `PRAGMA` vs. meta table for triggering execution, I see one advantage for the first. If we want to allow parameters for a call, we can easily use a vector of strings that defaults to be empty in the parser. Thus, the user can (i) omit providing a parameter if there is none, and (ii) provide multiple parameters. The downside, besides lacking conformance to the standard, is that the user has to know the plugin implementation. I would also note that we have other keywords that are not part of the standard, such as `IMPORT`, or `COPY` only being part of the Postgres dialect.
From my POV, having to know the plugin code is the most severe argument against a `PRAGMA` command and for the meta table. Though parameters could be provided via a string (similar to settings), the user would have to always provide an empty string if there is no parameter. Again, this requires knowledge of the plugin implementation, does not nicely adapt to functions with multiple parameters, and is kind of ugly when no parameters are required.

tl;dr: I'd say using a meta table and settings for parameters (default values, uniform query structure) maybe is the best solution for now. Just a minor note: do we want to let the meta table keep track of the call stack? This would make the history accessible even if plugins do not log their execution itself and stores execution and execution history at one place.

**Comment by Bouncner** [comment]
> From my POV, having to know the plugin code is the most severe argument against a `PRAGMA` command and for the meta table.

And do you have an opinion on the `std::map` vs. `EXPORT_USER_CALLABLE_FUNCTION` discussion?

> tl;dr: I'd say using a meta table and settings for parameters (default values, uniform query structure) maybe is the best solution for now. Just a minor note: do we want to let the meta table keep track of the call stack? This would make the history accessible even if plugins do not log their execution itself and stores execution and execution history at one place.

Sound reasonable and helpful. Should be basically for free, I guess. Would you store the call stack as a string in a `meta_exec_log` table with a timestamp or something different or more complex? Any ideas how to obtain the call stack? I have found `boost::stacktrace` and a bunch of platform-dependent stuff.

**Comment by Bensk1** [comment]
> Just a minor note: do we want to let the meta table keep track of the call stack? This would make the history accessible even if plugins do not log their execution itself and stores execution and execution history at one place.

Do you mean an actual call stack or just the history of user-triggered plugin function calls? These could be simply stored in the table itself. 

In general, how about moving the whole discussion to today's fortnightly? I think this could make the whole discussion more hyrisey,i.e., most efficient.

**Comment by dey4ss** [comment]
>Do you mean an actual call stack or just the history of user-triggered plugin function calls? These could be simply stored in the table itself.

Exactly, I thought about something like `std::vector<std::pair<std::string, std::string>>` within the meta table. If we return this list, it also preserves the semantics of inserting something into this table.

>In general, how about moving the whole discussion to today's fortnightly? I think this could make the whole discussion more hyrisey,i.e., most efficient.

+1

**Comment by Bensk1** [comment]
In today's Fortnightly we decided:

- to go for the alternative approach because it is more expressive and allows us to provide a list of user-callable functions to the user
- to use the `INSERT INTO` over the `PRAGMA` approach because it does not require parser adaption and is in general easier to implement. At the same time it does not prevent any future more powerful implementations
- to actually insert into the meta table which function was called to have a history of called functions.

File: src/test/lib/utils/meta_tables/meta_exec_table_test.cpp [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+  EXPECT_TRUE(sm.has_table("TableOfTestPlugin"));
+
+  SQLPipelineBuilder{
+      "INSERT INTO meta_exec (plugin_name, function_name) VALUES ('hyriseSecondTestPlugin', "
+      "'OurFreelyChoosableFunctionName')"}
+      .create_pipeline()
+      .get_result_table();
+  // The second test plugin creates the below table when the called function is executed
+  EXPECT_TRUE(sm.has_table("TableOfSecondTestPlugin"));
+}
```
Comment (Bensk1) [inline src/test/lib/utils/meta_tables/meta_exec_table_test.cpp:63]: These and the following tests kind of mirror some of the PluginManagerTests. It's pretty much the except that the SQLPipeline and everything that comes with it is in between. However, it's still nice to see everything working from end to end but we don't have to keep it.

File: src/lib/utils/plugin_manager.cpp [hunk 3]
```diff
@@ -60,21 +66,48 @@ void PluginManager::load_plugin(const std::filesystem::path& path) {
...
 
-void PluginManager::unload_plugin(const PluginName& name) {
-  auto plugin_iter = _plugins.find(name);
-  Assert(plugin_iter != _plugins.cend(), "Unloading plugin failed: A plugin with name " + name + " does not exist.");
+void PluginManager::exec_user_function(const PluginName& plugin_name, const PluginFunctionName& function_name) {
+  Assert(_user_executable_functions.count({plugin_name, function_name}) > 0,
+         "There is no " + function_name + " defined for plugin " + plugin_name + ".");
+
+  const auto user_executable_function = _user_executable_functions[{plugin_name, function_name}];
+  user_executable_function();
```
Comment (Bensk1) [inline src/lib/utils/plugin_manager.cpp:83]: This call is blocking. The called plugin could do some long running fancy optimization here which would be bad. Currently, it should be the plugin's responsibility to keep these functions short and, e.g., start a thread. We could also execute the function in a thread to avoid problems here. Also, what happens if `user_executable_function()` crashes?

File: src/test/lib/utils/lossless_predicate_cast_test.cpp [hunk 4]
```diff
@@ -18,7 +18,7 @@ TEST_F(LosslessPredicateCastTest, NextFloatTowards) {
   // Edge cases:
   EXPECT_EQ(next_float_towards(3.1, 3.1), std::nullopt);
 
-  // // Maximum double that can losslessly represented as a float and the next higher double
+  // Maximum double that can losslessly represented as a float and the next higher double
```
Comment (Bensk1) [inline src/test/lib/utils/lossless_predicate_cast_test.cpp:21]: This has nothing to do with plugins or this PR but I found the double comment while searching for that in the code.

Comment (benrobby) [reply src/test/lib/utils/lossless_predicate_cast_test.cpp:21]: 👍

File: src/test/lib/utils/meta_tables/meta_exec_table_test.cpp [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+
+TEST_F(MetaExecTest, CallNotCallableUserExecutableFunctions) {
+  auto& pm = Hyrise::get().plugin_manager;
+
+  // We have to manually rollback the transaction contexts below because otherwise their destruction would cause an
+  // exeception to be thrown. This is due to an assert in the TransactionContext's destructor checking for failed
+  // operators. See ~TransactionContext for details.
+
+  // Call non-existing function
```
Comment (benrobby) [inline src/test/lib/utils/meta_tables/meta_exec_table_test.cpp:72]: maybe `// Call non-existing plugin (with non-existing function)`?

**Review by benrobby — COMMENTED** [review]
@Bouncner invited me to also take a look since I'm also writing a plugin for my master thesis right now. When this is merged we could maybe add the functionality to the example plugin (https://github.com/hyrise/example_plugin).

File: src/lib/utils/plugin_manager.cpp [hunk 6]
```diff
@@ -60,21 +66,48 @@ void PluginManager::load_plugin(const std::filesystem::path& path) {
...
+
+  auto& log_manager = Hyrise::get().log_manager;
+  log_manager.add_message(
+      "PluginManager",
+      "Called user executable function `" + function_name + "` provided by plugin `" + plugin_name + "`.",
```
[the remark this answers is not in the document]
Comment (Bensk1) [reply src/lib/utils/plugin_manager.cpp:88]: First to the general discussion on how to handle long running functions. I talked to @Bouncner about that. We would leave it to the plugin to take care of long running functions and document that, e.g., in the AbstractPlugin.

Making users aware is a good idea. However, users might get a distorted picture if the blocking call of the PluginManager terminates quickly but the actual functionality executed by the plugin itself in a separate thread runs forever. I don't have a strong opinion here but I would not add wall clock information. Maybe we can get a third opinion @Bouncner?

**Comment by Bensk1** [comment]
Thanks for the review, I addressed the points you mentioned and added a comment to AbstractPlugin.

File: src/lib/utils/plugin_manager.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (Bouncner) [reply src/lib/utils/plugin_manager.cpp:88]: Hmm, I think extending the message by "(function returned after 17 ms)" can't hurt and might help in certain situations.
In case somebody thinks starting a thread in the function is a good idea, this person should be aware that one cannot rely on the runtime in the log message. What users are we talking about, when we talk about "users might get a distorted pictures"? I'd assume we'll never have the situation where the "user" executing the plugin has not been (at partially) involved in developing the plugin. I wouldn't see a problem in this case.

Or did I get you wrong?

File: src/lib/utils/plugin_manager.cpp [hunk 7]
```diff
@@ -60,21 +66,48 @@ void PluginManager::load_plugin(const std::filesystem::path& path) {
...
+  for (const auto& [function_name, function_pointer] : user_executable_functions) {
+    _user_executable_functions[{plugin_name, function_name}] = function_pointer;
+  }
 }
 
-void PluginManager::unload_plugin(const PluginName& name) {
-  auto plugin_iter = _plugins.find(name);
-  Assert(plugin_iter != _plugins.cend(), "Unloading plugin failed: A plugin with name " + name + " does not exist.");
+void PluginManager::exec_user_function(const PluginName& plugin_name, const PluginFunctionName& function_name) {
+  Assert(_user_executable_functions.count({plugin_name, function_name}) == 1,
```
Comment (Bouncner) [inline src/lib/utils/plugin_manager.cpp:79]: Minor: since that's the key there cannot be more than 1 and thus I'd favor contains().

File: src/lib/utils/plugin_manager.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (Bensk1) [reply src/lib/utils/plugin_manager.cpp:88]: I agree that it won't hurt. On the other hand, this is a user-triggered blocking function. Hence, if the function takes forever, the user will notice because the function will not return quickly. If the function takes 17ms the call returns immediately. On the console or via a client, the response typically even contains the execution time.

No, I think you understood me correctly. Plugin users will probably be involved in plugin development. However, plugins might be released as part of research papers where the user was not involved in the development. 

All in all, it's easy to add the timing information and it does not come with much overhead. However, I don't see much benefit of it. In particular because we want the PluginManager's blocking call to always finish really quickly, because the user is usually involved in the development, and we want the heavy lifting to be done inside the plugin itself.

**Comment by Bouncner** [comment]
Can't comment above. I don't know why.

> I agree that it won't hurt. On the other hand, this is a user-triggered blocking function. Hence, if the function takes forever, the user will notice because the function will not return quickly. If the function takes 17ms the call returns immediately. On the console or via a client, the response typically even contains the execution time.

Yeah, makes totally sense. I am fine without timing information. What about you @benrobby ?

**Comment by benrobby** [comment]
> Yeah, makes totally sense. I am fine without timing information. What about you @benrobby ?

Same for me 👍

File: src/plugins/CMakeLists.txt [hunk 8]
```diff
@@ -34,6 +34,7 @@ function(add_plugin)
 endfunction(add_plugin)
 
 add_plugin(NAME hyriseMvccDeletePlugin SRCS mvcc_delete_plugin.cpp mvcc_delete_plugin.hpp DEPS sqlparser magic_enum gtest)
+add_plugin(NAME hyriseSecondTestPlugin SRCS second_test_plugin.cpp second_test_plugin.hpp DEPS sqlparser)
```
[the remark this answers is not in the document]
Comment (Bensk1) [reply src/plugins/CMakeLists.txt:37]: Second one so that we can have better tests: testing name clashes of user callable functions originating from different plugins. 

Second before first because it is ordered alphabetically.

File: src/test/lib/utils/plugin_manager_test.cpp [hunk 9]
```diff
@@ -69,6 +69,83 @@ TEST_F(PluginManagerTest, LoadPluginAutomaticUnload) {
...
+  EXPECT_EQ(plugins.size(), 0u);
+  pm.load_plugin(build_dylib_path("libhyriseTestPlugin"));
+  pm.load_plugin(build_dylib_path("libhyriseSecondTestPlugin"));
+  EXPECT_EQ(plugins.size(), 2u);
+
+  {
+    auto user_executable_functions = pm.user_executable_functions();
+
+    EXPECT_EQ(user_executable_functions.size(), 3);
+    EXPECT_EQ(user_executable_functions.count({"hyriseSecondTestPlugin", "OurFreelyChoosableFunctionName"}), 1);
```
Comment (Bouncner) [inline src/test/lib/utils/plugin_manager_test.cpp:85]: 3* `.contains()`? clang-tidy (versions 13+) might complain.

**Review by Bouncner — APPROVED** [review]
I tiny thing. Otherwise, I thinks it's good to go.

File: src/test/lib/utils/plugin_manager_test.cpp (hunk 9 above)
Comment (Bensk1) [reply src/test/lib/utils/plugin_manager_test.cpp:85]: Makes sense. I think some time ago I always wanted to use contains but it was unsupported. Now it is supported and I don't use it :)

Comment (Bensk1) [reply src/test/lib/utils/plugin_manager_test.cpp:85]: I fixed some other occurrences along the way.
